### PR TITLE
[Theme] Speed up theme synchronization

### PR DESCRIPTION
**Pull request by frandiox**
I think this speeds up the dev command startup by a big margin:

- Make password check async
- Make checksum async while
- Upload file batches in parallel (relying on API throttler) instead of 1 by 1
- Delete files in parallel (relying on API throttler) instead of 1 by 1
- File deletion is made async so that the server can start before finishing

---

- I think this fixes a bug in file batching because we were not using the base64 size of certain files.
- Prevents re-reading files from disk because we can reuse the content that was read beforehand to calculate checksums.

---

- Still need to defer certain upload files. That might come in a different PR.
- To review, it might be easier commit by commit.

### CODE REVIEW

File: packages/theme/src/cli/services/dev.ts [hunk 1]
```diff
@@ -99,7 +96,12 @@ export async function dev(options: DevOptions) {
     },
   }
 
-  const server = await setupDevServer(options.theme, ctx)
+  const {server, renderProgress} = await setupDevServer(options.theme, ctx)
+
+  const storefrontPassword = await storefrontPasswordPromise
+  session.storefrontPassword = storefrontPassword
+
+  await renderProgress()
```
Comment (jamesmengo) [inline packages/theme/src/cli/services/dev.ts:104]: 👍🏻 great idea!

How do you feel about a rename to clarify that this is unrelated to the `renderLinks` call below?

Comment (frandiox) [reply packages/theme/src/cli/services/dev.ts]: Renamed to `renderThemeSyncProgress` 👍

**Review by jamesmengo — APPROVED** [review]
🎩 LGTM + code and works as expected 👍🏻

**Comment by frandiox** [comment]
@jamesmengo I've made further changes that speed up the startup a lot. Please give it another check 🙏

File: packages/theme/src/cli/utilities/theme-uploader.ts [hunk 2]
```diff
@@ -32,82 +33,171 @@ export async function uploadTheme(
...
+      title: titleGetter(),
+      task: async () => {
+        const result = await Promise.race([
+          promise,
+          new Promise((resolve) => setTimeout(() => resolve('timeout'), timeout)),
+        ])
+
+        if (result === 'timeout') {
+          addNextCheck()
+        }
```
Comment (frandiox) [inline packages/theme/src/cli/utilities/theme-uploader.ts:97]: This adds a new task when the timeout ends until `promise` is resolved. This way we update the title with a progress %

File: packages/theme/src/cli/utilities/theme-uploader.ts [hunk 3]
```diff
@@ -117,100 +207,26 @@ async function selectUploadableFiles(
...
-    }
-  })
-  return {
-    tasks,
-    updatedFileCount: runningFileCount,
-  }
-}
-
-async function createBatches(files: Checksum[], path: string): Promise<FileBatch[]> {
-  const fileSizes = await Promise.all(files.map((file) => fileSize(`${path}/${file.key}`)))
```
[the remark this answers is not in the document]
Comment (jamesmengo) [reply packages/theme/src/cli/utilities/theme-uploader.ts:207]: Thank you for cleaning this up!
This has been an area of the codebase I was not proud of and had been hoping to revisit for a while so I'm happy to see these improvements :)

File: packages/theme/src/cli/services/push.ts [hunk 4]
```diff
@@ -34,13 +34,21 @@ export async function push(theme: Theme, session: AdminSession, options: PushOpt
   const themeChecksums = await fetchChecksums(theme.id, session)
   const themeFileSystem = await mountThemeFileSystem(options.path)
 
-  const results = await uploadTheme(theme, session, themeChecksums, themeFileSystem, options)
+  const {uploadResults, renderThemeSyncProgress} = await uploadTheme(
```
Comment (jamesmengo) [inline packages/theme/src/cli/services/push.ts:37]: This method is hanging / not terminating when I 🎩 btw!

Comment (jamesmengo) [reply packages/theme/src/cli/services/push.ts:37]: I think [these changes](https://github.com/Shopify/cli/pull/4367/files#diff-d857e2b583dcee2197fc538eb4d35aabb5ce83cee18ebb970ece7854dda8ba07) will resolve that issue for us

**Review by jamesmengo — APPROVED** [review]
Left a comment [here](https://github.com/Shopify/cli/pull/4362/files#r1727760845) to make sure we're on the same page, but no changes needed on this particular changeset as the PR will resolve with [this](https://github.com/Shopify/cli/pull/4356/files)

**Comment by frandiox** [comment]
I've force-pushed to remove the last 2 commits and move them to this follow-up PR: https://github.com/Shopify/cli/pull/4370

I think this can be merged as is and will continue in that other PR after also mergin James' changes into main branch.

File: packages/theme/src/cli/utilities/theme-uploader.ts [hunk 5]
```diff
@@ -32,82 +33,183 @@ export async function uploadTheme(
...
+
+      return createBatches(fileType).map((batch) => {
+        return uploadBatch(batch, themeFileSystem, session, theme.id, uploadResults).then(() => {
+          progress.current += batch.length
+        })
+      })
+    }),
+  ).then(() => {
+    progress.current = progress.total
+  })
```
Comment (karreiro) [inline packages/theme/src/cli/utilities/theme-uploader.ts:200]: When we flat map the files, I wonder if we're exposing the logic of the order of files, as requests for Liquid sections and JSON templates are triggered simultaneously.

The files with dependencies we have are:

1. Liquid sections need to be uploaded first so they may be referenced by templates.
2. JSON sections need to be uploaded afterward so they can reference Liquid sections.
3. JSON templates should follow so they can reference sections.
4. Contextualized templates can be uploaded after templates.
5. Config files must be the last ones, but we need to upload config/settings_schema.json first, followed by config/settings_data.json.

The files with no dependencies we have are:

- The other Liquid files (for example, snippets).
- The other JSON files (for example, locales).
- The static assets.

We do have the `handleFailedUploads` function that retries 2 times if something goes wrong. However, given the variability in theme sizes (some themes have very large Liquid files, others have large JSON files), I'm concerned that if we don't guarantee the dependency order, the retry mechanism might not be enough to ensure successful uploads.

**Review by karreiro — COMMENTED** [review]
Thank you for this PR, @frandiox! This is awesome 🔥 It's 5x faster in my tests.

I've left one concern that might make the upload process a bit slower, but it would still be faster than the version we have on `main`. Still, I believe it's worth handling that concern to ensure we keep the uploader stable. Please let me know your thoughts about that :)

Thanks again for the amazing work!
